# Environmental and Surrounding Conditions at Locations of Serious and Fatal crashes in Christchurch between 2012 and 2022
### Authors: Kenneth Rosal(kro109), Kyle Hargraves(kha166), Miguel Montalla(mmo191), Xiaoshi Xie(xxi53)
### Date: 2022-10-21

Crash dataset containing nationwide crashes were downloaded from NZTA website(https://opendata-nzta.opendata.arcgis.com/datasets/NZTA::crash-analysis-system-cas-data-1/explore?location=-12.136558%2C0.000000%2C1.97) as a csv file. Then it was fitered with Christchurch region in Microsoft Excel and the result was exported to a new file called "Crash_data_chch.csv". This "Crash_data_chch.csv" was put in a public repository on Github (https://github.com/zelta1990/crash_data_nz). The source code for this project can be found in a repository called DATA422-Group-Project (https://github.com/zelta1990/DATA422-Group-Project). To run this notebook you shouldn't need to download or move any files. All you need to do is to uncomment the lines of install.packages to make sure all packages are installed on your machine.


The first step is to download the data from the repository via a URL. Then we inspect columns of the initial dataframe and filter the data to only have crashes happened after 2011. 

In [2]:
#install.packages("tidyverse")
#install.packages("readr")
library(readr)
library(tidyverse)
crash_df <-read_csv("https://raw.githubusercontent.com/zelta1990/crash_data_nz/main/Crash_data_chch.csv")
crash_df %>% head() #check initial dataset
crash_df %>% glimpse() #check column types
crash_df <- crash_df %>% filter(crashYear > 2011)  #only return records after 2012
crash_df <- crash_df  %>%  relocate(c(crashLocation1,crashLocation2), .after = Y) #move crashLocation 1 & 2 values next to co-ordinates for clarity #*****
head(crash_df,10) #*****

  

Rows: 51903 Columns: 72
-- Column specification --------------------------------------------------------
Delimiter: ","
chr (20): crashDirectionDescription, crashFinancialYear, crashLocation1, cra...
dbl (50): X, Y, OBJECTID, advisorySpeed, areaUnitID, bicycle, bridge, bus, c...
lgl  (2): crashRoadSideRoad, intersection

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


X       Y       OBJECTID advisorySpeed areaUnitID bicycle bridge bus
1 1569263 5177269  2       NA            595000     0       NA     0  
2 1554448 5181256  6       NA            587821     0        0     0  
3 1572260 5180996  8       NA            593300     0       NA     0  
4 1573557 5180808 16       NA            593600     0       NA     0  
5 1573407 5180049 22       NA            593600     0       NA     0  
6 1568675 5183354 27       NA            592100     0       NA     0  
  carStationWagon cliffBank <U+22EF>        train tree truck unknownVehicleType urban
1 2               NA        <U+22EF> NA    NA   0     0                  Urban
2 1                0        <U+22EF>  0     0   0     0                  Open 
3 2               NA        <U+22EF> NA    NA   0     0                  Urban
4 2               NA        <U+22EF> NA    NA   0     0                  Urban
5 1               NA        <U+22EF> NA    NA   0     0                  Urban
6 3               NA        <U+22EF> NA    NA   0     0                  Urban
  vanOrUtility vehicle waterRiver weatherA   weatherB
1 0            NA      NA         Fine       Null    
2 0             0       1         Light rain Null    
3 0            NA      NA         Fine       Null    
4 0            NA      NA         Fine       Null    
5 1            NA      NA         Heavy rain Null    
6 0            NA      NA         Fine       Null

Rows: 51,903
Columns: 72
$ X                         <dbl> 1569263, 1554448, 1572260, 1573557, 1573407,~
$ Y                         <dbl> 5177269, 5181256, 5180996, 5180808, 5180049,~
$ OBJECTID                  <dbl> 2, 6, 8, 16, 22, 27, 30, 36, 37, 38, 44, 46,~
$ advisorySpeed             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
$ areaUnitID                <dbl> 595000, 587821, 593300, 593600, 593600, 5921~
$ bicycle                   <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,~
$ bridge                    <dbl> NA, 0, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA~
$ bus                       <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,~
$ carStationWagon           <dbl> 2, 1, 2, 2, 1, 3, 2, 1, 2, 2, 2, 2, 1, 0, 1,~
$ cliffBank                 <dbl> NA, 0, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA~
$ crashDirectionDescription <chr> "East", "West", "North", "Null", "Null", "No~
$ crashFinancialYear        <chr> "1999/2000", "1999/2000", "2000/2001", "2000~
$ crashLocation

X       Y       crashLocation1    crashLocation2   OBJECTID advisorySpeed
1  1570633 5181084 COLOMBO ST        SALISBURY ST     124772   NA           
2  1571760 5179559 FITZGERALD AVENUE ST ASAPH ST      124800   NA           
3  1572679 5183607 MARSHLAND ROAD    JOY ST           124803   NA           
4  1569069 5185262 SH 74             MAIN NORTH ROAD  124814   NA           
5  1562376 5180519 SH 1S             BUCHANANS ROAD   125021   NA           
6  1567807 5185662 VEITCHES ROAD     BROCKHAM ST      125034   NA           
7  1565501 5179964 Z CPK COUNTDOWN   HANSONS LANE     125677   NA           
8  1566625 5184935 HAREWOOD ROAD     BISHOPDALE COURT 125727   NA           
9  1570752 5181272 BEALEY AVENUE     SHERBORNE ST     125820   NA           
10 1576291 5166958 MARINE DRIVE      NGAIO LANE       125824   NA           
   areaUnitID bicycle bridge bus <U+22EF>        train tree truck unknownVehicleType
1  591700     0       NA     0   <U+22EF> NA    NA   0     0                 
2  591500     0       NA     0   <U+22EF> NA    NA   0     0                 
3  592702     0        0     0   <U+22EF>  0     0   0     0                 
4  588102     0       NA     0   <U+22EF> NA    NA   0     0                 
5  587812     0       NA     0   <U+22EF> NA    NA   0     0                 
6  588402     0        0     0   <U+22EF>  0     1   0     0                 
7  587701     0       NA     0   <U+22EF> NA    NA   0     0                 
8  588700     0       NA     0   <U+22EF> NA    NA   0     0                 
9  591700     0       NA     0   <U+22EF> NA    NA   0     0                 
10 596502     0        0     0   <U+22EF>  0     0   0     0                 
   urban vanOrUtility vehicle waterRiver weatherA    weatherB
1  Urban 0            NA      NA         Light rain  Null    
2  Urban 0            NA      NA         Mist or Fog Null    
3  Urban 0             0       0         Mist or Fog Null    
4  Open  0            NA      NA         Fine        Null    
5  Urban 0            NA      NA         Fine        Null    
6  Urban 0             0       0         Light rain  Null    
7  Urban 0            NA      NA         Fine        Null    
8  Urban 0            NA      NA         Fine        Null    
9  Urban 1            NA      NA         Fine        Null    
10 Urban 0             0       0         Fine        Null

Now we begin to explore locations where serious or fatal crashes took place. We calculate the accident count for each location every year and rank the results only by count. We chose the combination of crashLocation1 and crashLocation2 instead of X and Y (NZGD2000 coordinates) as geographical reference for grouping because we are more interested in an area/intersection that has repeated accidents rather than a specific dot on the map. X and Y give too much precision, which is not helpful for us to find a geographical pattern on accident reoccurence.

In [3]:
#Calculate the count of accidents at one location for each year
count_df <- crash_df %>% 
                filter(crashSeverity == "Serious Crash" | crashSeverity == "Fatal Crash" ) %>%
                count(crashLocation1, crashLocation2, crashYear,sort=TRUE)  
head(count_df,10)

#Find how many times crashes reoccur or not at each combination of Location 1 and 2 #*****
count_df %>% count(n) %>% rename('Crash Occurence based on crashLoc 1 & crashLoc2' = n, n = nn)


crashLocation1                 crashLocation2  crashYear n
1  CHRISTCHURCH AKAROA ROAD       CATONS CV       2021      3
2  SH 74                          RADCLIFFE ROAD  2012      3
3  ALDWINS ROAD                   LINWOOD AVENUE  2019      2
4  ALDWINS ROAD                   MARCROFT ST     2016      2
5  BARRINGTON ST                  LINCOLN ROAD    2015      2
6  BLENHEIM ROAD                  CLARENCE ST     2017      2
7  BLENHEIM ROAD                  CLARENCE STREET 2019      2
8  BLENHEIM ROAD                  HANSONS LANE    2017      2
9  CHRISTCHURCH SOUTHERN MOTORWAY ANNEX UNDERPASS 2020      2
10 CLARENCE ST                    BLENHEIM ROAD   2016      2

Storing counts in `nn`, as `n` already present in input
i Use `name = "new_name"` to pick a new name.


Crash Occurence based on crashLoc 1 & crashLoc2 n   
1 1                                               1523
2 2                                                 52
3 3                                                  2

The above result shows that 52 locations had two accident occurence in one particular year and 2 locations had three accidents in one particular year. 

After we get the accident count for each location every year, we can rearrange the dataframe to find the locations with highest reoccurence in each year. We can also find if a certain year had more locations with accident reoccurences than other years.

In [5]:
#Rank each location by accident count for each year. Understand crash occurence per location for each year. #*****
all_year = list()
for(i in unique(crash_df$crashYear))
{
    year_df <- count_df %>% 
                    filter(crashYear == i ) %>% 
                    arrange(desc(n)) 
    all_year[[i]] <- year_df
}

all_year_df <- bind_rows(all_year)
all_year_df %>% arrange() %>% head()

#Using 2021 as an example
#all_year_df %>% filter(crashYear == '2021') 

#Summarise for each year; how many times crashes reoccur or not at each unique Location 1 and 2 combination. #*****
all_year_df_count <- all_year_df  %>% group_by(crashYear,n) %>% 
                    count() %>%
                    rename('Crash occurence per Loc 1 & Loc 2'= n, n = nn) 
               

all_year_df_count


crashLocation1 crashLocation2     crashYear n
1 SH 74          RADCLIFFE ROAD     2012      3
2 MERRIN ST      WITHELLS ROAD      2012      2
3 SH 75          KINLOCH ROAD S     2012      2
4 SH 76          COLOMBO ST         2012      2
5 BARBADOES ST   MOORHOUSE AVENUE   2012      1
6 BARRINGTON ST  BARRINGTON OFF WBD 2012      1

Storing counts in `nn`, as `n` already present in input
i Use `name = "new_name"` to pick a new name.


crashYear Crash occurence per Loc 1 & Loc 2 n  
1  2012      1                                 156
2  2012      2                                   3
3  2012      3                                   1
4  2013      1                                 162
5  2013      2                                   9
6  2014      1                                 168
7  2014      2                                   6
8  2015      1                                 162
9  2015      2                                   3
10 2016      1                                 149
11 2016      2                                   6
12 2017      1                                 143
13 2017      2                                   9
14 2018      1                                 139
15 2018      2                                   5
16 2019      1                                 142
17 2019      2                                   3
18 2020      1                                 104
19 2020      2                                   4
20 2021      1                                 117
21 2021      2                                   3
22 2021      3                                   1
23 2022      1                                  81
24 2022      2                                   1

Let's do a comparison using geographical references with a higher precision (involving X and Y) and see how many locations had accident occurences in each year. 

In [6]:
#Calculate the count of accidents at one location based on the unique combination of X,Y co-ordinates and crashLocation1 & crashLocation2. #*****
#This is done to confirm that each crash was generally recorded with its own unique position (X,Y location). 
#Helps us understand how crashes would be visualised (i.e. single points along road rather than stacked points in the same location). Need to show clustering of points in map as a result.
count_df2 <- crash_df %>% 
                filter(crashSeverity == "Serious Crash" | crashSeverity == "Fatal Crash" ) %>%
                count(X, Y,crashLocation1,crashLocation2,crashYear,sort=TRUE) 
head(count_df2)

#Rank each location by accident count for each year #*****
all_year = list()
for(i in unique(crash_df$crashYear))
{
    year_df <- count_df2 %>% 
                    filter(crashYear == i ) %>% 
                    arrange(desc(n)) 
    all_year[[i]] <- year_df
}
all_year_df2 <- bind_rows(all_year)

#Find how many times crashes reoccur or not at each combination of X, Y, crashLocation1 and crashLocation2 #*****
all_year_df2 %>% arrange(desc(n)) %>% count(n) %>% 
                rename('Crash occurence based on (X, Y, crashLoc1 & crashLoc2)'= n, n = nn) 


#Results show only 10 were reported to re-occur at the same exact X,Y(likely due to precision limits or error). #*****
#Thinking practically, this is minimal compared to all the crashes with unique X,Y position (1613)

X       Y       crashLocation1 crashLocation2    crashYear n
1 1560908 5178594 SH 1S          PARKER ST         2017      2
2 1561195 5184850 POUND ROAD     SAVILLS ROAD      2018      2
3 1565235 5186693 SH 1S          SAWYERS ARMS ROAD 2013      2
4 1567573 5180265 RICCARTON ROAD DIVISION ST       2018      2
5 1567930 5179478 CLARENCE ST    BLENHEIM ROAD     2016      2
6 1569448 5187955 SH 74          RADCLIFFE ROAD    2012      2

Storing counts in `nn`, as `n` already present in input
i Use `name = "new_name"` to pick a new name.


Crash occurence based on (X, Y, crashLoc1 & crashLoc2) n   
1 1                                                      1613
2 2                                                        10

Unsurprisingly, we got less locations with accident reoccurences. So let's stick to using the combination of crashLocation1 and crashLocation2 to find locations of interest. 

Is including crashYear for grouping a good idea to give us enough locations of interest? The results so far only give us 54 locations with accident reoccurence at most. What if we exclude crashYear when we do grouping and just do an accumulative count of accidents at each location over the 10 years then return the ones that had more than one accident? 

In [10]:
#Calculate the accumulative count of serious or fatal accidents at one location over ten years and return locations with count > 1 #*****
count_all_year_df <- crash_df %>% 
                filter(crashSeverity == "Serious Crash" | crashSeverity == "Fatal Crash" ) %>%
                count(crashLocation1, crashLocation2,sort=TRUE) 
count_all_year_df <- count_all_year_df %>%
                        filter(n > 1) #only return locations with accident count > 1
nrow(count_all_year_df)
count_all_year_df %>% head()



[1] 207

crashLocation1    crashLocation2 n
1 GEBBIES PASS ROAD SUMMIT ROAD    9
2 DYERS PASS ROAD   SUMMIT ROAD    7
3 SH 74             RADCLIFFE ROAD 7
4 DYERS PASS ROAD   CENTAURUS ROAD 6
5 HILLS ROAD        SHIRLEY ROAD   6
6 SH 74             MARSHLAND ROAD 6

Now we have 207 locations with accident reoccurence. Let's start working on these locations. First let's convert the NZGD2000 coordinates to WGS84 coordinates so that we can use them for visualisation later. Thanks to mkennedy's post on gis.stackexchange.com we got it a working function to convert the coordinates.  Information in this link https://spatialreference.org/ref/epsg/2193/ was also used to find an alternative working proj4string. 

In [7]:
#    Title: NZTM-WGS84(NZGD2000) converter
#    Author: mkennedy
#    Date: Feb 16, 2012 at 2:32
#    Availability: https://gis.stackexchange.com/questions/20389/converting-nzmg-or-nztm-to-latitude-longitude-for-use-with-r-map-library/20401#20401

#install.packages("proj4")
library(proj4)
get_lat_lon <- function(x,y)
{
    proj4string <- "+proj=tmerc +lat_0=0 +lon_0=173 +k=0.9996 +x_0=1600000 +y_0=10000000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"
    #proj4string <- "+proj=tmerc +lat_0=0.0 +lon_0=173.0 +k=0.9996 +x_0=1600000.0 +y_0=10000000.0 +datum=WGS84 +units=m"  # an alternative string that gives the same result
    # Source data
    xy <- data.frame(x, y)

    # Transformed data
    pj <- project(xy, proj4string, inverse=TRUE)
    latlon <- data.frame(lat=pj$y, lon=pj$x)
    return(latlon)
}
# an example
lat_lon <- get_lat_lon(1562170,5178795)
lat_lon


lat       lon     
1 -43.54283 172.5317

Now let's make a subset from the initial dataframe to only include locations of interest where crash severity was serious or fatal. Meanwhile we add lat and lon columns which were converted from X and Y using the get_lat_lon function above.

In [11]:
#Join count_all_year_df with original crash_df to return details for locations that had repeated accidents 
target_df <- inner_join(crash_df, count_all_year_df %>% 
                select(-3), by = c('crashLocation1', 'crashLocation2')) %>%
                filter(crashSeverity == "Serious Crash" | crashSeverity == "Fatal Crash" ) %>% 
                mutate(lat = get_lat_lon(X,Y)$lat, lon=get_lat_lon(X,Y)$lon)

dim(target_df)           
target_df %>% relocate(c(crashLocation1,crashLocation2,lat,lon),.after = Y)  %>%  head() #Shift lat,lon and crash location vectors for clarity

[1] 508  74

X       Y       crashLocation1    crashLocation2    lat       lon     
1 1572933 5184220 LAKE TERRACE ROAD FAIRWAY DRIVE     -43.49446 172.6652
2 1570500 5178327 SH 76             COLOMBO ST        -43.54742 172.6348
3 1574466 5184474 BURWOOD ROAD      LAKEWOOD DRIVE    -43.49222 172.6842
4 1563005 5181023 SH 73             CUTTS ROAD        -43.52282 172.5422
5 1568560 5178680 LINCOLN ROAD      CLARENCE ST SOUTH -43.54417 172.6108
6 1569953 5174972 HACKTHORNE ROAD   STAMBRIDGE PLACE  -43.57761 172.6279
  OBJECTID advisorySpeed areaUnitID bicycle <U+22EF>        train tree truck
1 129062   NA            592702     0       <U+22EF>  0     0   0    
2 129113   NA            594600     0       <U+22EF> NA    NA   0    
3 129153   NA            590505     0       <U+22EF> NA    NA   0    
4 129372   NA            587812     0       <U+22EF> NA    NA   0    
5 130297   NA            594700     1       <U+22EF> NA    NA   0    
6 130609   NA            591101     1       <U+22EF> NA    NA   0    
  unknownVehicleType urban vanOrUtility vehicle waterRiver weatherA weatherB
1 0                  Urban 0             0       0         Fine     Null    
2 0                  Urban 1            NA      NA         Fine     Null    
3 0                  Urban 0            NA      NA         Fine     Null    
4 0                  Urban 0            NA      NA         Fine     Null    
5 0                  Urban 0            NA      NA         Fine     Null    
6 0                  Urban 0            NA      NA         Fine     Null

Now we have got a dataframe with 508 rows. Remember each combination of crashLocation1 and crashLocation2 may have more than one corresponding X and Y. That's why we didn't get only 207 rows. This dataframe has got 74 columns but do we really need this many columns for our study? How about we do a frequency count and percentage calculation for each column and see if they actually have got useful high-frequency values? For this calculation we are only concerned about columns that are related to environmental and surrounding conditions hence we will exclude irrelevant columns before we do the calculation.

In [14]:
#Calculate frequency for factor levels in columns. The result was used as a reference to determine which columns should be used for layers in the map
col_freq_df <- target_df %>% select(-c(X,Y,OBJECTID, areaUnitID,carStationWagon,crashSHDescription,crashFinancialYear,crashDirectionDescription,crashLocation1,crashLocation2,crashRoadSideRoad,crashSeverity,
                                        crashYear,debris,directionRoleDescription,ditch,fatalCount,fence,holiday,houseOrBuilding,intersection,meshblockId,minorInjuryCount,
                                        objectThrownOrDropped,otherObject,otherVehicleType,overBank,parkedVehicle,moped,pedestrian,phoneBoxEtc,postOrPole,region,roadworks,roadCharacter,
                                        schoolBus,seriousInjuryCount,suv,temporarySpeedLimit,taxi,tlaId,tlaName,urban,unknownVehicleType,vehicle,vanOrUtility,lat,lon))
all_columns = list()
for(i in 1:ncol(col_freq_df))
{
    df <- col_freq_df[i]
    new_df <- df %>%
                group_by(df[1]) %>%
                tally() %>%
                mutate(percent=round(100*n/sum(n))) %>%
                arrange(desc(percent)) 
    colnames(new_df)[1] <- "value"  
    colnames(new_df)[2] <- "count"         
    new_df <- cbind(name = colnames(col_freq_df[i]), new_df)
    new_df$value <- as.character(new_df$value)
    all_columns[[i]] <- new_df
}
all_col_freq_df <- bind_rows(all_columns)
#all_col_freq_df  #*****
all_columns   #show frequency table for each column individually #**** 
#all_col_freq_df %>%  filter(name == 'trafficControl')  #filter table for variable of interest


[[1]]
           name value count percent
1 advisorySpeed  <NA>   489      96
2 advisorySpeed    35     8       2
3 advisorySpeed    45     4       1
4 advisorySpeed    65     3       1
5 advisorySpeed    20     1       0
6 advisorySpeed    25     2       0
7 advisorySpeed    30     1       0

[[2]]
     name value count percent
1 bicycle     0   406      80
2 bicycle     1    98      19
3 bicycle     2     3       1
4 bicycle     4     1       0

[[3]]
    name value count percent
1 bridge  <NA>   389      77
2 bridge     0   119      23

[[4]]
  name value count percent
1  bus     0   495      97
2  bus     1    13       3

[[5]]
       name value count percent
1 cliffBank  <NA>   389      77
2 cliffBank     0   101      20
3 cliffBank     1    17       3
4 cliffBank     2     1       0

[[6]]
      name     value count percent
1 flatHill      Flat   442      87
2 flatHill Hill Road    65      13
3 flatHill      Null     1       0

[[7]]
       name value count percent
1 guardRail  <NA>   389      77
2 guardRail     0   113      22
3 guardRail     1     6       1

[[8]]
  name value count percent
1 kerb  <NA>   389      77
2 kerb     0   117      23
3 kerb     1     2       0

[[9]]
   name      value count percent
1 light Bright sun   241      47
2 light       Dark   132      26
3 light   Overcast   115      23
4 light   Twilight    19       4
5 light    Unknown     1       0

[[10]]
        name value count percent
1 motorcycle     0   420      83
2 motorcycle     1    83      16
3 motorcycle     2     5       1

[[11]]
           name value count percent
1 NumberOfLanes     2   361      71
2 NumberOfLanes     4    73      14
3 NumberOfLanes     3    35       7
4 NumberOfLanes     6    18       4
5 NumberOfLanes     5    10       2
6 NumberOfLanes     1     6       1
7 NumberOfLanes  <NA>     4       1
8 NumberOfLanes     7     1       0

[[12]]
      name    value count percent
1 roadLane    2-way   459      90
2 roadLane    1-way    47       9
3 roadLane Off road     2       0

[[13]]
         name    value count percent
1 roadSurface   Sealed   506     100
2 roadSurface Unsealed     2       0

[[14]]
         name value count percent
1 slipOrFlood  <NA>   389      77
2 slipOrFlood     0   119      23

[[15]]
        name value count percent
1 speedLimit    50   261      51
2 speedLimit    60    88      17
3 speedLimit   100    55      11
4 speedLimit    80    50      10
5 speedLimit    70    45       9
6 speedLimit    30     5       1
7 speedLimit    40     2       0
8 speedLimit  <NA>     2       0

[[16]]
         name value count percent
1 strayAnimal  <NA>   389      77
2 strayAnimal     0   119      23

[[17]]
         name value count percent
1 streetLight  Null   262      52
2 streetLight    On   125      25
3 streetLight   Off    77      15
4 streetLight  None    44       9

[[18]]
            name           value count percent
1 trafficControl         Unknown   157      31
2 trafficControl Traffic Signals   131      26
3 trafficControl             Nil    98      19
4 trafficControl        Give way    89      18
5 trafficControl            Stop    33       6

[[19]]
           name value count percent
1 trafficIsland  <NA>   389      77
2 trafficIsland     0   112      22
3 trafficIsland     1     7       1

[[20]]
         name value count percent
1 trafficSign  <NA>   389      77
2 trafficSign     0   108      21
3 trafficSign     1    11       2

[[21]]
   name value count percent
1 train  <NA>   389      77
2 train     0   119      23

[[22]]
  name value count percent
1 tree  <NA>   389      77
2 tree     0   100      20
3 tree     1    19       4

[[23]]
   name value count percent
1 truck     0   469      92
2 truck     1    34       7
3 truck     2     5       1

[[24]]
        name value count percent
1 waterRiver  <NA>   389      77
2 waterRiver     0   118      23
3 waterRiver     1     1       0

[[25]]
      name       value count percent
1 weatherA        Fine   447      88
2 weatherA  Light rain  

Looks like some columns have NA or 0 accounted for a significant percentage such as train, strayAnimal. We probably don't need columns like those. 

We are also interested in whether a certain condition has experienced changes over years at a location. We can use unique() function to find the unique values in one specific column for each location. If the length of the vector returned by unique() function is larger than 1 then it suggests there might have been changes for that specific condition at that specific location. We explore this possibility with advisorySpeed and trafficControl as they are very likely to change over years.

In [16]:
#Find number of changes to variables over 10 years. These will be the number of unique values for a specific variable for each crash location 1 and 2
#Exploratory analysis to track number of unique values of variables at crash locations(traffic Control, speed limit, etc)

#Number of changes for advisory speed
advsspeed_changes <- target_df  %>% group_by(crashLocation1,crashLocation2)  %>%  summarise(num_advsspeed = length(unique(advisorySpeed)))
arrange(advsspeed_changes , desc(advsspeed_changes$num_advsspeed))  %>%  head(n=10)

#Check manually for top result for advisory speed
check_df <- target_df  %>% filter(crashLocation1 == 'GOVERNORS BAY TEDDINGTON' & crashLocation2 == 'ALLANDALE LANE')  %>%  select(crashLocation1,crashLocation2,advisorySpeed,crashSeverity)
check_df

#Number of changes for traffic Control
trafficControl_changes <- target_df  %>% group_by(crashLocation1,crashLocation2)  %>%  summarise(num_trafficControl = length(unique(trafficControl)))
arrange(trafficControl_changes , desc(trafficControl_changes$num_trafficControl))  %>%  head(n=10)

#Check manually for top result for traffic control
check_df2 <- target_df  %>% filter(crashLocation1 == 'BLENHEIM ROAD' & crashLocation2 == 'WHARENUI ROAD')  %>%  select(crashLocation1,crashLocation2,trafficControl)
check_df2

`summarise()` has grouped output by 'crashLocation1'. You can override using
the `.groups` argument.


crashLocation1           crashLocation2     num_advsspeed
1  GOVERNORS BAY TEDDINGTON ALLANDALE LANE     3            
2  CHRISTCHURCH AKAROA ROAD CATONS CV          2            
3  DYERS PASS ROAD          CENTAURUS ROAD     2            
4  DYERS PASS ROAD          HACKTHORNE ROAD    2            
5  DYERS PASS ROAD          PENTRE TERRACE     2            
6  DYERS PASS ROAD          SUMMIT ROAD        2            
7  EVANS PASS ROAD          OCEAN VIEW TERRACE 2            
8  GOVERNORS BAY ROAD       SANDY BEACH ROAD   2            
9  MARSHS ROAD              FOUNTAINS ROAD     2            
10 SH 75                    HILLTOP HOTEL      2

crashLocation1           crashLocation2 advisorySpeed crashSeverity
1 GOVERNORS BAY TEDDINGTON ALLANDALE LANE 30            Serious Crash
2 GOVERNORS BAY TEDDINGTON ALLANDALE LANE NA            Serious Crash
3 GOVERNORS BAY TEDDINGTON ALLANDALE LANE 35            Serious Crash

`summarise()` has grouped output by 'crashLocation1'. You can override using
the `.groups` argument.


crashLocation1    crashLocation2   num_trafficControl
1  BLENHEIM ROAD     WHARENUI ROAD    3                 
2  GEBBIES PASS ROAD SUMMIT ROAD      3                 
3  HILLS ROAD        SHIRLEY ROAD     3                 
4  MARSHLAND ROAD    PRESTONS ROAD    3                 
5  ROYDVALE AVENUE   MEMORIAL AVENUE  3                 
6  SH 73             BLENHEIM ROAD    3                 
7  AVONHEAD ROAD     RON GUTHREY ROAD 2                 
8  AVONSIDE DRIVE    TRENT ST         2                 
9  AWATEA ROAD       SPRINGS ROAD     2                 
10 BARRINGTON ST     COBHAM ST        2

crashLocation1 crashLocation2 trafficControl 
1 BLENHEIM ROAD  WHARENUI ROAD  Nil            
2 BLENHEIM ROAD  WHARENUI ROAD  Traffic Signals
3 BLENHEIM ROAD  WHARENUI ROAD  Unknown

We have done quite a bit of exploratory analysis. Now let's wrangle the dataframe to have only columns that we are interested in. 
We might want to convert columns apart from lat, lon and crashYear to char type so we can make them factors, a preparation for later visualisation. We also want to convert NA and Null to a more readable value such as "Unknown". Removing NAs in this case might not be a good idea because it would bump up other values' percentage, creating a false impression that some conditions account for large proportions of the accident count.

In [23]:
# End wide data frame - Based on lat,long co-ordinates and relevant variables deemed usedul to crashes that can be addressed by Council. #*****
d <- as.data.frame(target_df %>%         
        select(lat,lon,crashLocation1, crashLocation2,crashYear,trafficControl,light,advisorySpeed,bicycle,cliffBank,flatHill,guardRail,motorcycle,NumberOfLanes,
        roadLane,speedLimit,streetLight,trafficIsland,weatherA,weatherB)) 
d <- replace(d, is.na(d), "Unknown")  
d[ d == "Null" ] <- "Unknown"
d[-c(1,2,5)] <- sapply(d[-c(1,2,5)],as.character) #convert all layer columns to char type 
head(d)



lat       lon      crashLocation1    crashLocation2    crashYear
1 -43.49446 172.6652 LAKE TERRACE ROAD FAIRWAY DRIVE     2012     
2 -43.54742 172.6348 SH 76             COLOMBO ST        2013     
3 -43.49222 172.6842 BURWOOD ROAD      LAKEWOOD DRIVE    2013     
4 -43.52282 172.5422 SH 73             CUTTS ROAD        2013     
5 -43.54417 172.6108 LINCOLN ROAD      CLARENCE ST SOUTH 2013     
6 -43.57761 172.6279 HACKTHORNE ROAD   STAMBRIDGE PLACE  2013     
  trafficControl  light      advisorySpeed bicycle cliffBank flatHill 
1 Give way        Bright sun Unknown       0       0         Flat     
2 Unknown         Bright sun Unknown       0       Unknown   Flat     
3 Traffic Signals Bright sun Unknown       0       Unknown   Flat     
4 Give way        Bright sun Unknown       0       Unknown   Flat     
5 Give way        Bright sun Unknown       1       Unknown   Flat     
6 Nil             Bright sun Unknown       1       Unknown   Hill Road
  guardRail motorcycle NumberOfLanes roadLane speedLimit streetLight
1 0         0          2             2-way    50         Unknown    
2 Unknown   0          2             1-way    60         Unknown    
3 Unknown   1          3             2-way    50         Unknown    
4 Unknown   1          4             2-way    60         Unknown    
5 Unknown   0          2             2-way    50         Unknown    
6 Unknown   0          2             2-way    50         Unknown    
  trafficIsland weatherA weatherB
1 0             Fine     Unknown 
2 Unknown       Fine     Unknown 
3 Unknown       Fine     Unknown 
4 Unknown       Fine     Unknown 
5 Unknown       Fine     Unknown 
6 Unknown       Fine     Unknown

Now that we have got dataframe sorted out, it's time to visualise it. We will use leaflet package and R shiny to build an interactive map. Apart from using the datafram we have got from the last step, we will also be using the csv file webscrapped from NZTA traffic cam API by Julia code (check nzta_api_Julia.ipnyb)for camera markers. The cameras.csv we will be using is a copy we stored in the crash_data_nz repo on github (https://github.com/zelta1990/crash_data_nz). On running the code below you would be able to see the interactive map on your localhost with the port specified in the output. You can also see a deployed version that is hosted on Rshiny server by visiting https://mmo191.shinyapps.io/deploy422/

In [27]:
#Plotting interactive map using leaflet
#install.packages("shiny")
#install.packages("RColorBrewer")
#install.packages("sp")
#install.packages("leaflet.extras2")
#install.packages("ggplot2")
library(sp)
library(leaflet)
library(leaflet.extras2)
library(shiny)
library(RColorBrewer)
library(ggplot2)


camera_d <- read.csv("https://raw.githubusercontent.com/zelta1990/crash_data_nz/main/cameras.csv") %>%    # read web cameras details from data scraped from NZTA traffic cam API using Julia
            filter(region == "Christchurch")
camera_d$lat <- as.numeric(camera_d$lat)                                                                                 
camera.icon <- makeIcon("https://img.icons8.com/arcade/64/000000/camera--v1.png",iconWidth = 20, iconHeight = 20)

ui <- fillPage(  
                 titlePanel(""),
                 mainPanel(tags$style(type = "text/css", "#m {height: calc(100vh - 40px) !important;}"),
                           leafletOutput("m"), width = 8),                                                   # Put map in main panel 
                 sidebarPanel(                                                                                           # Put year and layer filter in side bar
                              width = 4,
                              sliderInput("year",
                                          "Year",
                                          min=min(d$crashYear),
                                          max=max(d$crashYear),
                                          value=range(d$crashYear),
                                          step=1, sep=""),
                              checkboxInput("checkbox", "Show clusters", value = TRUE),                                  # A toggle for displaying markers as clusters
                              selectInput("layer","Select a layer: ", c("trafficControl","speedLimit","advisorySpeed","light","streetLight","bicycle","motorcycle","cliffBank","flatHill","guardRail","NumberOfLanes",
                                          "roadLane","trafficIsland","weatherA","weatherB")), # Add layer selector
                              tags$a(href="https://opendata-nzta.opendata.arcgis.com/pages/cas-data-field-descriptions", " Click here for layer descriptions",target="_blank"),
                              plotOutput("pie"),                                                                         # Draw pie chart for frequency count of factor levels of the selected layer column for the selected period of time
                              downloadButton("downloadData", "Download")                                                 # Add a button to download the filtered years' data for the layer columns
                              ))
server <- function(input, output, session){
  url <- a("Layer Description", href="https://opendata-nzta.opendata.arcgis.com/pages/cas-data-field-descriptions")
    output$tab <- renderUI({
      tagList("URL link:", url)
    })
  filteredData <- reactive({                                                                                             # filter data by selected years using slider
    d[d$crashYear >= input$year[1] & d$crashYear <= input$year[2],]                                                       
  })
  d_colour <- colorFactor( palette="RdYlBu", domain=NULL, na.color="transparent")                                        # color palette for legends
  output$pie <- renderPlot(                                                                                               
    ggplot(count(d,d[, 'trafficControl']), aes(x="", y=unlist(count(d,d[, input$layer])[2]), fill=unlist(count(d,d[, input$layer])[1]))) +   #Draw piechart
              geom_bar(stat="identity", width=1) + coord_polar("y", start=0) +labs(fill = input$layer)+xlab("")+ylab("") +
              geom_label(aes(label = unlist(count(d,d[, input$layer])[2])),
                        color = "#000000",
                        position = position_stack(vjust = 0.5),
                        show.legend = FALSE) +
              scale_fill_brewer(palette="RdYlBu")
  )
  output$downloadData <- downloadHandler (                                                                                # Download filtered data on pressing button
      filename = function(){
          paste("filtered_data_by_year","csv",sep=".")
            },
      
          content = function(file){
           write.csv(filteredData(), file)
            })
  output$m <- renderLeaflet({                                                                                             # draw map
  
    leaflet(d) %>% 
      setView( lat=-43.5, lng=172.6 , zoom=10) %>%                                                                        
      addTiles() %>%    
      addMarkers(lng = camera_d$lon, lat = camera_d$lat,                                                                  #add markers for cameras 
                icon = camera.icon,
                popup = paste0("<img src = ", camera_d$imageUrl, " width = 300 >"))%>%
      addCircleMarkers(~d$lon, ~d$lat,                                                                                    #add markers for locations
        fillColor = ~d_colour(d$trafficControl), fillOpacity = 0.9, color="#af4343", radius=8, stroke=FALSE,
        popup=paste0("<br>trafficControl:",d$trafficControl,"<br>lat:", d$lat, "<br>lon:", d$lon,"<br>year:", d$crashYear,"<br>location1:", d$crashLocation1,"<br>location2:", d$crashLocation2 ),
        clusterOptions = switch(input$checkbox==TRUE,markerClusterOptions(spiderfyDistanceMultiplier=1.5),NULL)
      ) %>%
      addLegend(pal=d_colour, values=~trafficControl, opacity=0.9, title = "trafficControl", position = "bottomright")
    })
  
  observe({                                                                                                               # Rearrange display on input changes
    output$pie <- renderPlot(
    ggplot(count(filteredData(),filteredData()[, input$layer]), aes(x="", y=unlist(count(filteredData(),filteredData()[, input$layer])[2]), fill=unlist(count(filteredData(),filteredData()[, input$layer])[1]))) + 
              geom_bar(stat="identity", width=1) + coord_polar("y", start=0) +labs(fill = input$layer)+xlab("")+ylab("") +
              geom_label(aes(label = unlist(count(filteredData(),filteredData()[, input$layer])[2])),
                color = "#000000",
                position = position_stack(vjust = 0.5),
                show.legend = FALSE) +
                scale_fill_brewer(palette="RdYlBu")
  )
    leafletProxy("m", data=filteredData()) %>% 
      clearControls() %>%                                                                                                 # Update legend
      clearMarkers() %>%                                                                                
      clearMarkerClusters()%>%
      addMarkers(lng = camera_d$lon, lat = camera_d$lat,                                                                  # redraw markers
            icon = camera.icon,
            popup = paste0("<img src = ", camera_d$imageUrl, " width = 300 >"))%>%
      addCircleMarkers(~lon, ~lat, 
        fillColor = ~d_colour(filteredData()[,input$layer]), fillOpacity = 0.9, color="#af4343", radius=8, stroke=FALSE,
        popup=~paste0("<br>",input$layer,":",filteredData()[,input$layer],"<br>lat:", lat, "<br>lon:", lon,"<br>year:", crashYear ),
        clusterOptions = switch(input$checkbox==TRUE,markerClusterOptions(spiderfyDistanceMultiplier=1.5),NULL)           # Toggle between cluster mode and dots mode
       ) %>%
       addLegend(pal=d_colour, values=filteredData()[,input$layer], opacity=0.9, title = input$layer, position = "bottomright")
  } )
}

shinyApp(ui, server) 


Listening on http://127.0.0.1:3611

